recursion/compiler/src/circuit/compiler.rs

```diff
@@ -0,0 +1,653 @@
+            Entry::Vacant(entry) => panic!("expected entry in fp_to_addr: {entry:?}"),
+            Entry::Occupied(entry) => {
+                // This is a read, so we increment the mult.
+                match self.addr_to_mult.get_mut(entry.get()) {
+                    Some(mult) => *mult += F::one(),
+                    None => panic!("expected entry in addr_mult: {entry:?}"),
+                }
+                *entry.into_mut()
+            }
+        }
+    }
+
+    /// Map `fp` to a fresh address and initialize the mult to 0.
+    ///
+    /// Ensures that `fp` has not already been written to.
+    pub fn write_fp(&mut self, fp: i32) -> Address<F> {
+        match self.fp_to_addr.entry(fp) {
+            Entry::Vacant(entry) => {
+                let addr = Self::alloc(&mut self.next_addr);
+                // This is a write, so we set the mult to zero.
+                if let Some(x) = self.addr_to_mult.insert(addr, F::zero()) {
+                    panic!("unexpected entry in addr_to_mult: {x:?}");
+                }
+                *entry.insert(addr)
+            }
+            Entry::Occupied(entry) => panic!("unexpected entry in fp_to_addr: {entry:?}"),
+        }
+    }
+
+    /// Increment the existing `mult` associated with `addr`.
+    ///
+    /// Ensures that `addr` has already been assigned a `mult`.
+    pub fn read_addr(&mut self, addr: Address<F>) -> F {
+        match self.addr_to_mult.entry(addr) {
+            Entry::Vacant(entry) => panic!("expected entry in addr_to_mult: {entry:?}"),
+            Entry::Occupied(entry) => {
+                // This is a read, so we increment the mult.
+                let mult = entry.into_mut();
+                *mult += F::one();
+                *mult
+            }
+        }
+    }
+
+    /// Associate a `mult` of zero with `addr`.
+    ///
+    /// Ensures that `addr` has not already been written to.
+    pub fn write_addr(&mut self, addr: Address<F>) -> F {
+        match self.addr_to_mult.entry(addr) {
+            Entry::Vacant(entry) => *entry.insert(F::zero()),
+            Entry::Occupied(entry) => panic!("unexpected entry in addr_to_mult: {entry:?}"),
+        }
+    }
+
+    /// Read the base field constant.
+    ///
+    /// Increments the mult, first creating an entry if it does not yet exist.
+    pub fn read_const_f(&mut self, f: F) -> Address<F> {
+        self.consts_f
+            .entry(f)
+            .and_modify(|(_, x)| *x += F::one())
+            .or_insert_with(|| (Self::alloc(&mut self.next_addr), F::one()))
+            .0
+    }
+
+    /// Read the base field constant.
+    ///
+    /// Increments the mult, first creating an entry if it does not yet exist.
+    pub fn read_const_ef(&mut self, ef: EF) -> Address<F> {
+        self.consts_ef
+            .entry(ef)
+            .and_modify(|(_, x)| *x += F::one())
+            .or_insert_with(|| (Self::alloc(&mut self.next_addr), F::one()))
+            .0
+    }
+
+    // INSTRUCTION HELPERS
```

Comment:
stylistically would just remove, if you really want to seperate them out can use trait or different impl block